### PR TITLE
SWP-695 - Add customers and refunds

### PR DESCRIPTION
**Pull request by lukasz-kowal**

### CODE REVIEW

File: src/Entity/Customer.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace PodPoint\Payments\Entity;
```
Comment (LeeBrooks3) [inline src/Entity/Customer.php:3]: don't mind the naming of `entities` or `models` (plural preferred) but I feel both of these could be misconstrued to be related to a database which is why I didn't add or feel a need for a namespace.

maybe worth having @AndreiaMatiasSilva's second opinion on this?

File: src/Entity/Customer.php [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+
+    public function getUid(): string
+    {
+        return $this->uid;
+    }
+
+    public function getEmail(): string
+    {
+        return $this->email;
+    }
```
Comment (LeeBrooks3) [inline src/Entity/Customer.php:31]: I think as this is a POPO (plain old PHP object) it is fine to have public properties without getter/setters

File: src/Exception.php [hunk 3]
```diff
@@ -1,14 +0,0 @@
...
-
-class Exception extends \Exception
-{
-    /**
-     * @param \Throwable|null $previous
-     */
-    public function __construct(\Throwable $previous = null)
-    {
-        parent::__construct('Failed to create payment due to provider exception', 1, $previous);
-    }
```
Comment (LeeBrooks3) [inline src/Exception.php:13]: I have an idea on how we can approach exceptions in this project, I'll create a PR from this branch to show you as I think that's the easiest way

File: src/Providers/Stripe/Service.php [hunk 4]
```diff
@@ -24,25 +29,40 @@ public function __construct(string $key)
...
-     * @throws Exception
+     * @throws PaymentException
      * @throws StripeException
      */
-    public function create(string $token, int $amount, string $currency = 'GBP'): Payment
-    {
+    public function create(
+        string $token,
+        int $amount,
+        ?string $customerId = null,
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:45]: nullable types with `?` 👌  nice - haven't seen before, but I'd prefer it to always be a string if it is passed, we should never really be passing `null` as an argument here IMO

File: src/Providers/Stripe/Exception.php [hunk 5]
```diff
@@ -2,22 +2,20 @@
...
+     * @var mixed
      */
     private $response;
 
     /**
-     * @param PaymentIntent $response
+     * @param mixed $response
      * @param \Throwable|null $previous
      */
-    public function __construct(PaymentIntent $response, \Throwable $previous = null)
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Exception.php:20]: it was quite good to have this typehinted - I think a different exception should be used if this is going to be a different type

File: src/Exceptions/PaymentException.php [hunk 6]
```diff
@@ -0,0 +1,15 @@
+<?php
+
+namespace PodPoint\Payments\Exceptions;
+
+class PaymentException extends \Exception
```
Comment (LeeBrooks3) [inline src/Exceptions/PaymentException.php:5]: not keen on the naming on this, we are already within the `\Payments` namespace (the whole library is) so I don't think the word `Payment` is necessary here

File: src/Providers/Stripe/Service.php [hunk 7]
```diff
@@ -24,25 +29,40 @@ public function __construct(string $key)
...
+        ?string $description = null,
+        array $metadata = [],
+        string $currency = 'GBP'
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:48]: not sure of the ordering here - I think maybe `currency` should still come after amount

File: src/Providers/Stripe/Service.php [hunk 8]
```diff
@@ -24,25 +29,40 @@ public function __construct(string $key)
...
+     * @throws PaymentException
      * @throws StripeException
      */
-    public function create(string $token, int $amount, string $currency = 'GBP'): Payment
-    {
+    public function create(
+        string $token,
+        int $amount,
+        ?string $customerId = null,
+        ?string $description = null,
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:46]: is description not part of the `metadata`, or could it be?

File: src/Providers/Stripe/Service.php (hunk 4 above)
Comment (LeeBrooks3) [reply src/Providers/Stripe/Service.php:45]: also when is this passed? on second payments?

File: src/Providers/Stripe/Service.php [hunk 9]
```diff
@@ -77,4 +100,74 @@ public function update(string $uid): Payment
...
+     * @param string $paymentMethod
+     * @param string|null $description
+     *
+     * @return Customer
+     */
+    public function createCustomer(string $email, string $paymentMethod, ?string $description = null): Customer
+    {
+        $customer = StripeCustomer::create([
+            'description' => $description,
+            'email'       => $email,
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:117]: we don't do arrow alignment

File: src/Providers/Stripe/Service.php [hunk 10]
```diff
@@ -77,4 +100,74 @@ public function update(string $uid): Payment
...
+    /**
+     * Creates Stripe customer.
+     *
+     * @param string $email
+     * @param string $paymentMethod
+     * @param string|null $description
+     *
+     * @return Customer
+     */
+    public function createCustomer(string $email, string $paymentMethod, ?string $description = null): Customer
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:113]: I think I agree with @adteulade that this belongs in another service class

Comment (LeeBrooks3) [reply src/Providers/Stripe/Service.php:113]: again I feel description isn't necessary / qualifies as meta-data, what is this used for currently?

File: src/Providers/Stripe/Service.php [hunk 11]
```diff
@@ -77,4 +100,74 @@ public function update(string $uid): Payment
...
+    {
+        $customer = StripeCustomer::create([
+            'description' => $description,
+            'email'       => $email,
+            'payment_method' => $paymentMethod,
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:118]: is this a necessary argument? and is it a token the same as it is for a payment? if so I would call it `token`.

File: src/Providers/Stripe/Service.php [hunk 12]
```diff
@@ -77,4 +100,74 @@ public function update(string $uid): Payment
...
+     * @param string|null $reason
+     * @param array|null $metadata
+     *
+     * @return Refund
+     *
+     * @throws PaymentException
+     * @throws StripeException
+     */
+    public function refund(
+        string $intentId,
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:138]: we're trying to be generic here so I would call this `uid` or `paymentUid`

File: src/Providers/Stripe/Service.php [hunk 13]
```diff
@@ -77,4 +100,74 @@ public function update(string $uid): Payment
...
+     * @param int|null $amount
+     * @param string|null $reason
+     * @param array|null $metadata
+     *
+     * @return Refund
+     *
+     * @throws PaymentException
+     * @throws StripeException
+     */
+    public function refund(
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:137]: also think this should belong in it's own service

File: src/Providers/Stripe/Service.php [hunk 14]
```diff
@@ -77,4 +100,74 @@ public function update(string $uid): Payment
 
         return new Payment($response->id, $response->currency, $response->amount, $response->created);
     }
+
+    /**
+     * Creates Stripe customer.
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:105]: `Tries create a customer using the Stripe SDK.`

File: src/Providers/Stripe/Service.php [hunk 15]
```diff
@@ -77,4 +100,74 @@ public function update(string $uid): Payment
...
+            'description' => $description,
+            'email'       => $email,
+            'payment_method' => $paymentMethod,
+        ]);
+
+        return new Customer($customer->id, $customer->email);
+    }
+
+    /**
+     * Creates Refund for provided Payment Intent id.
```
Comment (LeeBrooks3) [inline src/Providers/Stripe/Service.php:125]: `Tries refund the given payment using the Stripe SDK.`

File: src/Service.php [hunk 16]
```diff
@@ -24,8 +41,39 @@ public function create(string $token, int $amount, string $currency = 'GBP'): Pa
...
      * @return Payment
      *
-     * @throws Exception
+     * @throws PaymentException
      * @throws StripeException
      */
     public function update(string $uid): Payment;
+
+    /**
+     * Creates Stripe customer.
```
Comment (LeeBrooks3) [inline src/Service.php:50]: `Creates a customer.`

File: src/Service.php [hunk 17]
```diff
@@ -24,8 +41,39 @@ public function create(string $token, int $amount, string $currency = 'GBP'): Pa
...
+     * @param string $email
+     * @param string $paymentMethod
+     * @param string|null $description
+     *
+     * @return Customer
+     */
+    public function createCustomer(string $email, string $paymentMethod, ?string $description = null): Customer;
+
+    /**
+     * Creates Refund for provided Payment Intent id.
```
Comment (LeeBrooks3) [inline src/Service.php:61]: `Refunds the given payment.`

File: src/Entity/Customer.php (hunk 1 above)
Comment (lukasz-kowal) [reply src/Entity/Customer.php:3]: I don't mind either way, they are kind of Value Objects only at the moment, but they don't need to be in the dedicated folder.

File: src/Exception.php (hunk 3 above)
Comment (lukasz-kowal) [reply src/Exception.php:13]: cool

**Pull request closed without merging**